src/transformStateWithClones.js

```diff
@@ -7,7 +7,44 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const HISTORY = [];
+  let currentState = { ...state };
+
+  for (const ACTION of actions) {
+    switch (ACTION.type) {
+      case 'addProperties':
+        currentState = addProperties(currentState, ACTION.extraData);
+        break;
+
+      case 'removeProperties':
+        currentState = removeProperties(currentState, ACTION.keysToRemove);
+        break;
+
+      case `clear`:
+        currentState = {};
+        break;
+    }
+
+    HISTORY.push({ ...currentState });
+  }
+
+  return HISTORY;
+}
+
+function addProperties(state, extraData) {
+  const STATE_COPY = { ...state };
+
+  return Object.assign(STATE_COPY, extraData);
+}
+
+function removeProperties(state, keysToRemove) {
+  const STATE_COPY = { ...state };
```

Comment:
```suggestion
function addProperties(state, extraData) {

  return Object.assign(state, extraData);
}
```
you already make copy in 11 line